### PR TITLE
Add an Ai goal flag to afterburn hard

### LINKED ISSUE #5303: FR: Force afterburn on AI orders
As per discussion with Asteroth, this is a request for two additional arguments to AI orders.

The first arg is true/false bool for whether or not the AI uses afterburners to close distance to a target during an order, such as waypoints or destroy target.

The second is a defined radius around the target in which the AI should not use afterburners while executing the order.

This is a balance critical active mod request.

### PR DESCRIPTION
**Pull request by Baezon**
Allows modders to more or less force AI to afterburn as hard as possible, either by simply adding or short-cutting existing afterburning behavior. Not as a matter of difficulty or anything, but mission choreography. They still should afterburn intelligently though, this only makes them hustle to get near their goal, and they won't burn so hard they can't properly complete their goal, whether it's attacking or defending or whatever, their behavior once they get here is basically unchanged.

~~Additionally, this removes some superfluous checks for having afterburners or whether they are locked, `afterburner_start()` already checks both of these so they just muck up the code.~~ Since it makes different decisions based on the status of the afterburners they are not superfluous >.>

Closes #5303

### CODE REVIEW

File: code/ai/aicode.cpp [hunk 1]
```diff
@@ -12529,7 +12586,7 @@ int ai_formation()
 
 	ship_info *sip = &Ship_info[shipp->ship_info_index];
 	bool ab_allowed = false;
-	if ((sip->flags[Ship::Info_Flags::Afterburner]) && !(shipp->flags[Ship::Ship_Flags::Afterburner_locked]) && (aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use])) {
+	if ((aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use]) || ai_willing_to_afterburn_hard(aip)) {
```
Comment (BMagnu) [inline code/ai/aicode.cpp:12589]: This removes the check for locked afterburners.
Is this intentional?

File: code/ai/aicode.cpp [hunk 2]
```diff
@@ -10734,8 +10784,11 @@ void ai_guard()
 			compute_desired_rvec(&rvec, &goal_point, &Pl_objp->pos);
 			ai_turn_towards_vector(&goal_point, Pl_objp, nullptr, nullptr, 0.0f, 0, &rvec);
 
-			if ((aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use]) && !(shipp->flags[Ship::Ship_Flags::Afterburner_locked]) && (accel_scale * (0.25f + dist_to_goal_point/700.0f) > 0.8f)) {
-				if (ai_maybe_fire_afterburner(Pl_objp, aip)) {
```
Comment (BMagnu) [inline code/ai/aicode.cpp:10738]: here as well

File: code/ai/aicode.cpp [hunk 3]
```diff
@@ -10611,16 +10655,22 @@ void ai_big_guard()
 		ai_turn_towards_vector(&goal_pt, Pl_objp, nullptr, nullptr, 0.0f, 0);
 		accelerate_ship(aip, 1.0f);
 
-		if ((aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use]) && !(shipp->flags[Ship::Ship_Flags::Afterburner_locked]) && (cur_guard_rad > 1.1f * max_guard_dist)) {
+
+		bool free_ab_use = aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use];
+		if ((free_ab_use || ai_willing_to_afterburn_hard(aip)) && (cur_guard_rad > 1.1f * max_guard_dist)) {
```
Comment (BMagnu) [inline code/ai/aicode.cpp:10660]: here as well

File: code/ai/aicode.cpp [hunk 4]
```diff
@@ -4685,7 +4715,9 @@ void ai_fly_to_target_position(vec3d* target_pos, bool* pl_done_p=NULL, bool* pl
 			ab_allowed = false;
 	}
 
-	if (!(sip->flags[Ship::Info_Flags::Afterburner]) || (shipp->flags[Ship::Ship_Flags::Afterburner_locked]) || !(aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use])) {
+	if (!(aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || 
+		aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use] || 
+		ai_willing_to_afterburn_hard(aip))) {
```
Comment (BMagnu) [inline code/ai/aicode.cpp:4720]: here as well

File: code/ai/aibig.cpp [hunk 5]
```diff
@@ -849,7 +850,9 @@ void ai_big_chase_attack(ai_info *aip, ship_info *sip, vec3d *enemy_pos, float d
 			} else {
 				accelerate_ship(aip, accel);
 
-				if ((aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use]) && !(shipp->flags[Ship::Ship_Flags::Afterburner_locked]) && (accel > 0.95f)) {
+				if (ai_willing_to_afterburn_hard(aip) && accel > 0.99f) {
+					ai_afterburn_hard(Pl_objp, aip);
+				} else if ((aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use]) && !(shipp->flags[Ship::Ship_Flags::Afterburner_locked]) && (accel > 0.95f)) {
```
Comment (BMagnu) [inline code/ai/aibig.cpp:855]: But here interestingly not.

File: code/ai/aicode.cpp (hunk 1 above)
Comment (BMagnu) [reply code/ai/aicode.cpp:12589]: Just read the comment, duh.

File: code/ai/aibig.cpp (hunk 5 above)
Comment (Baezon) [reply code/ai/aibig.cpp:855]: Actually... on second thought, these often aren't superfluous...

It makes *different decisions* based on whether the afterburners are locked or not. It's trying and failing to afterburn vs not even attempting to afterburn and doing something else instead... so I will have to amend these regardless.

File: code/ai/aicode.cpp [hunk 6]
```diff
@@ -12529,7 +12594,7 @@ int ai_formation()
 
 	ship_info *sip = &Ship_info[shipp->ship_info_index];
 	bool ab_allowed = false;
-	if ((sip->flags[Ship::Info_Flags::Afterburner]) && !(shipp->flags[Ship::Ship_Flags::Afterburner_locked]) && (aip->ai_flags[AI::AI_Flags::Free_afterburner_use] || aip->ai_profile_flags[AI::Profile_Flags::Free_afterburner_use])) {
```
Comment (BMagnu) [inline code/ai/aicode.cpp:12532]: Well, if we want to keep the AB lock checks, it should probably also return here

**Review by BMagnu — APPROVED** [review]

**Review by Goober5000 — CHANGES_REQUESTED** [review]
One discrepancy; the rest looks ok

**Review by Goober5000 — APPROVED** [review]

**Pull request merged**